extension/persistence/relational-jdbc/src/main/java/org/apache/polaris/extension/persistence/relational/jdbc/DatasourceOperations.java

```diff
@@ -173,23 +189,105 @@ public int executeUpdate(String query) throws SQLException {
+          try (Connection connection = borrowConnection()) {
+            boolean autoCommit = connection.getAutoCommit();
+            boolean success = false;
+            connection.setAutoCommit(false);
+            try {
+              try {
+                try (Statement statement = connection.createStatement()) {
+                  success = callback.execute(statement);
+                }
+              } finally {
+                if (success) {
+                  connection.commit();
+                } else {
+                  connection.rollback();
+                }
+              }
+            } finally {
+              connection.setAutoCommit(autoCommit);
+            }
+          }
+          return null;
+        });
+  }
+
+  private boolean isRetryable(SQLException e) {
+    String sqlState = e.getSQLState();
+
+    if (sqlState != null) {
+      return sqlState.equals(SERIALIZATION_FAILURE_SQL_CODE); // Serialization failure
+    }
+
+    // Additionally, one might check for specific error messages or other conditions
+    return e.getMessage().toLowerCase(Locale.ROOT).contains("connection refused")
+        || e.getMessage().toLowerCase(Locale.ROOT).contains("connection reset");
+  }
+
+  // TODO: consider refactoring to use a retry library, inorder to have fair retries
+  // and more knobs for tuning retry pattern.
+  @VisibleForTesting
+  <T> T withRetries(Operation<T> operation) throws SQLException {
+    int attempts = 0;
+    // maximum number of retries.
+    int maxAttempts = relationalJdbcConfiguration.maxRetries().orElse(1);
+    // How long we should try, since the first attempt.
+    long maxDuration = relationalJdbcConfiguration.maxDurationInMs().orElse(5000L);
+    // How long to wait before first failure.
+    long delay = relationalJdbcConfiguration.initialDelayInMs().orElse(100L);
+
+    // maximum time we will retry till.
+    long maxRetryTime = clock.millis() + maxDuration;
+
+    while (attempts < maxAttempts) {
       try {
-        try (Statement statement = connection.createStatement()) {
-          success = callback.execute(statement);
-        }
-      } finally {
-        if (success) {
-          connection.commit();
+        return operation.execute();
+      } catch (SQLException | RuntimeException e) {
+        SQLException sqlException;
+        if (e instanceof RuntimeException) {
+          // Handle Exceptions from ResultSet Iterator consumer, as it throws a RTE, ignore RTE from
+          // the transactions.
+          if (e.getCause() instanceof SQLException
+              && !(e instanceof EntityAlreadyExistsException)) {
+            sqlException = (SQLException) e.getCause();
+          } else {
+            throw e;
+          }
         } else {
-          connection.rollback();
+          sqlException = (SQLException) e;
         }
-        connection.setAutoCommit(autoCommit);
+
+        attempts++;
+        long timeLeft = Math.max((maxRetryTime - clock.millis()), 0L);
```

Comment:
There's a rare but still existent problem here:
The code is using the _wall clock_, which can go backwards and forwards at any time. Not so much of an issue in _proper_ server setups (most NTP default setups), but even proper NTP setups can jump.
What you really want here is a strictly monotonically increasing clock (and no, System.nanoTime() alone is not a good candidate).

Reply:
`System.nanoTime()` is supposed to provide accurate _intervals_ (differences) when accessed from the same thread (as in our case here), so it should be fine if we want to switch from stopping retried based on a wall clock moment-in-time to counting time we spent in the retry loop from start to finish.

Reply:
@dimas-b that's only true if the thread doesn't migrate to another CPU - sadly

Reply:
The javadoc talks about our use case here as a supported use case :shrug: